pkg/controller/yurthub/crsapprover.go

```diff
@@ -0,0 +1,219 @@
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package yurthub
+
+import (
+	"context"
+	"crypto/x509"
+	"encoding/pem"
+	"fmt"
+	"time"
+
+	certificates "k8s.io/api/certificates/v1beta1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/apimachinery/pkg/util/wait"
+	certinformer "k8s.io/client-go/informers/certificates/v1beta1"
+	certv1beta1 "k8s.io/client-go/informers/certificates/v1beta1"
+	"k8s.io/client-go/kubernetes"
+	typev1beta1 "k8s.io/client-go/kubernetes/typed/certificates/v1beta1"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/util/workqueue"
+	"k8s.io/klog/v2"
+
+	"github.com/openyurtio/openyurt/pkg/projectinfo"
+	"github.com/openyurtio/openyurt/pkg/yurthub/pki/certmanager"
+)
+
+const (
+	YurtHubCSRApproverThreadiness = 2
+)
+
+// YurtHubCSRApprover is the controller that auto approve all
+// yurthub related CSR
+type YurtHubCSRApprover struct {
+	csrInformer certv1beta1.CertificateSigningRequestInformer
+	csrClient   typev1beta1.CertificateSigningRequestInterface
+	workqueue   workqueue.RateLimitingInterface
+}
+
+// Run starts the YurtHubCSRApprover
+func (yca *YurtHubCSRApprover) Run(threadiness int, stopCh <-chan struct{}) {
+	defer runtime.HandleCrash()
+	defer yca.workqueue.ShutDown()
+	klog.Info("starting the crsapprover")
+	if !cache.WaitForCacheSync(stopCh,
+		yca.csrInformer.Informer().HasSynced) {
+		klog.Error("sync csr timeout")
+		return
+	}
+	for i := 0; i < threadiness; i++ {
+		go wait.Until(yca.runWorker, time.Second, stopCh)
+	}
+	<-stopCh
+	klog.Info("stoping the csrapprover")
+}
+
+func (yca *YurtHubCSRApprover) runWorker() {
+	for yca.processNextItem() {
+	}
+}
+
+func (yca *YurtHubCSRApprover) processNextItem() bool {
+	key, quit := yca.workqueue.Get()
+	if quit {
+		return false
+	}
+	csrName, ok := key.(string)
+	if !ok {
+		yca.workqueue.Forget(key)
+		runtime.HandleError(
+			fmt.Errorf("expected string in workqueue but got %#v", key))
+		return true
+	}
+	defer yca.workqueue.Done(key)
+
+	csr, err := yca.csrInformer.Lister().Get(csrName)
+	if err != nil {
+		runtime.HandleError(err)
+		if !apierrors.IsNotFound(err) {
+			yca.workqueue.AddRateLimited(key)
+		}
+		return true
+	}
+
+	if err := approveYurtHubCSR(csr, yca.csrClient); err != nil {
+		runtime.HandleError(err)
+		enqueueObj(yca.workqueue, csr)
+		return true
+	}
+
+	return true
+}
+
+func enqueueObj(wq workqueue.RateLimitingInterface, obj interface{}) {
+	key, err := cache.MetaNamespaceKeyFunc(obj)
+	if err != nil {
+		runtime.HandleError(err)
+		return
+	}
+	wq.AddRateLimited(key)
```

Comment:
In order to reduce the overhead of csr controller, before add into worker queue, we need to verify the object, like the object is csr, or csr object is yurthub csr, or csr is approved or denied.

you can reference the following link: https://github.com/openyurtio/openyurt/blob/master/pkg/yurttunnel/pki/certmanager/csrapprover.go#L105